File: android/app/src/main/java/com/noop/analytics/StressSignalQuality.kt
package com.noop.analytics

data class StressSignalSample(
    val timestamp: Long,
    val value: Double?,
    val durationSeconds: Long? = null,
    val source: String,
    val provenance: String? = null,
    val quality: Quality,
    val isMissing: Boolean = false,
    val activityContext: ActivityContext = ActivityContext.UNKNOWN,
) {
    enum class Quality { VALID, REJECTED, UNKNOWN }
    enum class ActivityContext { UNKNOWN, RESTING, ACTIVE }
}

enum class StressQualityRejectionReason(val wire: String) {
    INVALID_WINDOW("invalid_window"),
    NO_VALID_SAMPLES("no_valid_samples"),
    INSUFFICIENT_TEMPORAL_COVERAGE("insufficient_temporal_coverage"),
    MAXIMUM_GAP_EXCEEDED("maximum_gap_exceeded"),
    BASELINE_NOT_READY("baseline_not_ready"),
}

enum class StressQualityDiagnostic(val wire: String) {
    DUPLICATE_TIMESTAMPS("duplicate_timestamps"),
    OUT_OF_ORDER_INPUT("out_of_order_input"),
    REJECTED_SAMPLES("rejected_samples"),
}

data class StressQualityConfiguration(
    val version: String,
    val minimumCoverageFraction: Double,
    val minimumValidDurationSeconds: Long,
    val maximumGapSeconds: Long,
    val maximumCreditedIntervalSeconds: Long,
    val nominalSampleDurationSeconds: Long,
) {
    companion object {
        val DAYTIME_V1 = StressQualityConfiguration(
            version = "stress-temporal-v1",
            minimumCoverageFraction = 300.0 / 3_600.0,
            minimumValidDurationSeconds = 300L,
            maximumGapSeconds = 15 * 60L,
            maximumCreditedIntervalSeconds = 60L,
            nominalSampleDurationSeconds = 1L,
        )
    }
}

data class StressQualityDecision(
    val accepted: Boolean,
    val coveragePercentage: Double,
    val maximumGapSeconds: Long,
    val validDurationSeconds: Long,
    val rejectedDurationSeconds: Long,
    val windowDurationSeconds: Long,
    val baselineReady: Boolean,
    val rejectionReasons: List<StressQualityRejectionReason>,
    val diagnostics: List<StressQualityDiagnostic>,
    val validSampleCount: Int,
    val rejectedSampleCount: Int,
    val duplicateTimestampCount: Int,
    val configurationVersion: String,
) {
    fun withBaselineReadiness(ready: Boolean): StressQualityDecision {
        val reasons = rejectionReasons.filterNot {
            it == StressQualityRejectionReason.BASELINE_NOT_READY
        }.toMutableList()
        if (!ready) reasons += StressQualityRejectionReason.BASELINE_NOT_READY
        return copy(accepted = reasons.isEmpty(), baselineReady = ready, rejectionReasons = reasons)
    }
}

object StressTemporalQuality {
    private data class CanonicalSample(val timestamp: Long, val value: Double, val durationSeconds: Long?)
    private data class Interval(val start: Long, val endExclusive: Long)

    fun evaluate(
        samples: List<StressSignalSample>,
        windowStart: Long,
        windowEnd: Long,
        baselineReady: Boolean,
        configuration: StressQualityConfiguration = StressQualityConfiguration.DAYTIME_V1,
    ): StressQualityDecision {
        val validWindow = windowEnd > windowStart &&
            !(windowStart < 0L && windowEnd > Long.MAX_VALUE + windowStart)
        val windowDuration = if (validWindow) windowEnd - windowStart else 0L
        val inWindow = samples.filter { it.timestamp >= windowStart && it.timestamp < windowEnd }
        val outOfOrder = inWindow.zipWithNext().any { (a, b) -> a.timestamp > b.timestamp }
        val duplicateCount = inWindow.size - inWindow.map { it.timestamp }.toSet().size
        val canonical = canonicalValidSamples(inWindow)
        val invalid = inWindow.filterNot(::isValid)
        val validIntervals = representedIntervals(canonical, windowStart, windowEnd, configuration)
        val rejectedIntervals = invalid.mapNotNull { sample ->
            if (windowDuration <= 0L) return@mapNotNull null
            val duration = minOf(
                sample.durationSeconds?.takeIf { it > 0L }
                    ?: configuration.nominalSampleDurationSeconds,
                configuration.maximumCreditedIntervalSeconds,
            )
            val end = minOf(windowEnd, safeAdd(sample.timestamp, duration.coerceAtLeast(0L)))
            if (end > sample.timestamp) Interval(sample.timestamp, end) else null
        }
        val validDuration = unionDuration(validIntervals)
        val rejectedDuration = unionDuration(rejectedIntervals)
        val maxGap = if (windowDuration > 0L) maximumGap(validIntervals, windowStart, windowEnd) else 0L
        val coverage = if (windowDuration > 0L) validDuration.toDouble() / windowDuration else 0.0

        val reasons = buildList {
            if (windowDuration <= 0L) add(StressQualityRejectionReason.INVALID_WINDOW)
            if (canonical.isEmpty()) add(StressQualityRejectionReason.NO_VALID_SAMPLES)
            if (windowDuration > 0L && (validDuration < configuration.minimumValidDurationSeconds ||
                    coverage + 1e-12 < configuration.minimumCoverageFraction)) {
                add(StressQualityRejectionReason.INSUFFICIENT_TEMPORAL_COVERAGE)
            }
            if (windowDuration > 0L && maxGap > configuration.maximumGapSeconds) {
                add(StressQualityRejectionReason.MAXIMUM_GAP_EXCEEDED)
            }
            if (!baselineReady) add(StressQualityRejectionReason.BASELINE_NOT_READY)
        }
        val diagnostics = buildList {
            if (duplicateCount > 0) add(StressQualityDiagnostic.DUPLICATE_TIMESTAMPS)
            if (outOfOrder) add(StressQualityDiagnostic.OUT_OF_ORDER_INPUT)
            if (invalid.isNotEmpty()) add(StressQualityDiagnostic.REJECTED_SAMPLES)
        }
        return StressQualityDecision(
            accepted = reasons.isEmpty(),
            coveragePercentage = coverage * 100.0,
            maximumGapSeconds = maxGap,
            validDurationSeconds = validDuration,
            rejectedDurationSeconds = rejectedDuration,
            windowDurationSeconds = windowDuration,
            baselineReady = baselineReady,
            rejectionReasons = reasons,
            diagnostics = diagnostics,
            validSampleCount = canonical.size,
            rejectedSampleCount = invalid.size,
            duplicateTimestampCount = duplicateCount,
            configurationVersion = configuration.version,
        )
    }

    fun canonicalMean(
        samples: List<StressSignalSample>, windowStart: Long, windowEnd: Long,
        configuration: StressQualityConfiguration = StressQualityConfiguration.DAYTIME_V1,
    ): Double? {
        val canonical = canonicalValidSamples(
            samples.filter { it.timestamp >= windowStart && it.timestamp < windowEnd },
        )
        var weighted = 0.0
        var duration = 0L
        for ((index, sample) in canonical.withIndex()) {
            val credit = minOf(sample.durationSeconds ?: configuration.nominalSampleDurationSeconds,
                configuration.maximumCreditedIntervalSeconds)
            val next = if (index + 1 < canonical.size) canonical[index + 1].timestamp else windowEnd
            val end = minOf(windowEnd, next, safeAdd(sample.timestamp, credit.coerceAtLeast(0L)))
            val seconds = (end - maxOf(windowStart, sample.timestamp)).coerceAtLeast(0L)
            weighted += sample.value * seconds
            duration += seconds
        }
        return if (duration > 0L) weighted / duration else null
    }

    private fun isValid(sample: StressSignalSample): Boolean =
        sample.quality == StressSignalSample.Quality.VALID && !sample.isMissing &&
            sample.value?.isFinite() == true && (sample.durationSeconds == null || sample.durationSeconds > 0L)

    private fun canonicalValidSamples(samples: List<StressSignalSample>): List<CanonicalSample> =
        samples.filter(::isValid).groupBy { it.timestamp }.toSortedMap().map { (timestamp, group) ->
            val values = group.mapNotNull { it.value }.sorted()
            CanonicalSample(timestamp, values.sum() / values.size, group.mapNotNull { it.durationSeconds }.maxOrNull())
        }

    private fun representedIntervals(
        samples: List<CanonicalSample>,
        windowStart: Long,
        windowEnd: Long,
        configuration: StressQualityConfiguration,
    ): List<Interval> {
        if (windowEnd <= windowStart) return emptyList()
        return samples.mapIndexedNotNull { index, sample ->
            val duration = minOf(
                sample.durationSeconds ?: configuration.nominalSampleDurationSeconds,
                configuration.maximumCreditedIntervalSeconds,
            )
            val start = maxOf(windowStart, sample.timestamp)
            val next = if (index + 1 < samples.size) samples[index + 1].timestamp else windowEnd
            val end = minOf(windowEnd, next, safeAdd(sample.timestamp, duration.coerceAtLeast(0L)))
            if (end > start) Interval(start, end) else null
        }
    }

    private fun unionDuration(intervals: List<Interval>): Long {
        val ordered = intervals.sortedBy { it.start }
        if (ordered.isEmpty()) return 0L
        var current = ordered.first()
        var total = 0L
        for (interval in ordered.drop(1)) {
            if (interval.start <= current.endExclusive) {
                current = Interval(current.start, maxOf(current.endExclusive, interval.endExclusive))
            } else {
                total += current.endExclusive - current.start
                current = interval
            }
        }
        return total + current.endExclusive - current.start
    }

    private fun maximumGap(intervals: List<Interval>, windowStart: Long, windowEnd: Long): Long {
        if (windowEnd <= windowStart) return 0L
        val ordered = intervals.sortedBy { it.start }
        if (ordered.isEmpty()) return windowEnd - windowStart
        var current = ordered.first()
        var largest = (current.start - windowStart).coerceAtLeast(0L)
        for (interval in ordered.drop(1)) {
            if (interval.start > current.endExclusive) {
                largest = maxOf(largest, interval.start - current.endExclusive)
                current = interval
            } else if (interval.endExclusive > current.endExclusive) {
                current = Interval(current.start, interval.endExclusive)
            }
        }
        return maxOf(largest, windowEnd - current.endExclusive)
    }

    private fun safeAdd(value: Long, positiveDelta: Long): Long =
        if (value > Long.MAX_VALUE - positiveDelta) Long.MAX_VALUE else value + positiveDelta
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/StressSignalQuality.swift
import Foundation

/// One source-normalized observation presented to the stress quality gate. Source-specific decoders
/// populate provenance and missingness; the quality engine deliberately ignores provenance when deciding
/// acceptance so equivalent imported, native, and replay samples produce identical results.
public struct StressSignalSample: Codable, Equatable, Sendable {
    public enum Quality: String, Codable, Sendable {
        case valid
        case rejected
        case unknown
    }

    public enum ActivityContext: String, Codable, Sendable {
        case unknown
        case resting
        case active
    }

    public let timestamp: Int
    public let value: Double?
    public let durationSeconds: Int?
    public let source: String
    public let provenance: String?
    public let quality: Quality
    public let isMissing: Bool
    public let activityContext: ActivityContext

    public init(timestamp: Int,
                value: Double?,
                durationSeconds: Int? = nil,
                source: String,
                provenance: String? = nil,
                quality: Quality,
                isMissing: Bool = false,
                activityContext: ActivityContext = .unknown) {
        self.timestamp = timestamp
        self.value = value
        self.durationSeconds = durationSeconds
        self.source = source
        self.provenance = provenance
        self.quality = quality
        self.isMissing = isMissing
        self.activityContext = activityContext
    }
}

public enum StressQualityRejectionReason: String, Codable, Sendable, CaseIterable {
    case invalidWindow = "invalid_window"
    case noValidSamples = "no_valid_samples"
    case insufficientTemporalCoverage = "insufficient_temporal_coverage"
    case maximumGapExceeded = "maximum_gap_exceeded"
    case baselineNotReady = "baseline_not_ready"
}

public enum StressQualityDiagnostic: String, Codable, Sendable, CaseIterable {
    case duplicateTimestamps = "duplicate_timestamps"
    case outOfOrderInput = "out_of_order_input"
    case rejectedSamples = "rejected_samples"
}

/// Versioned, engineering-only signal requirements. These are not physiological coefficients.
public struct StressQualityConfiguration: Codable, Equatable, Sendable {
    public let version: String
    public let minimumCoverageFraction: Double
    public let minimumValidDurationSeconds: Int
    public let maximumGapSeconds: Int
    public let maximumCreditedIntervalSeconds: Int
    public let nominalSampleDurationSeconds: Int

    public init(version: String,
                minimumCoverageFraction: Double,
                minimumValidDurationSeconds: Int,
                maximumGapSeconds: Int,
                maximumCreditedIntervalSeconds: Int,
                nominalSampleDurationSeconds: Int) {
        self.version = version
        self.minimumCoverageFraction = minimumCoverageFraction
        self.minimumValidDurationSeconds = minimumValidDurationSeconds
        self.maximumGapSeconds = maximumGapSeconds
        self.maximumCreditedIntervalSeconds = maximumCreditedIntervalSeconds
        self.nominalSampleDurationSeconds = nominalSampleDurationSeconds
    }

    /// Keeps the previous five-minutes-per-hour evidence floor (300 seconds / 3,600 seconds), but unlike
    /// the old count rule it rejects a five-minute burst followed by a long hole. Explicit decoder-supplied
    /// durations may represent at most one minute; an observation with no duration earns only the nominal
    /// one second, so the gate never fabricates coverage from an unknown sampling interval.
    public static let daytimeV1 = StressQualityConfiguration(
        version: "stress-temporal-v1",
        minimumCoverageFraction: 300.0 / 3_600.0,
        minimumValidDurationSeconds: 300,
        maximumGapSeconds: 15 * 60,
        maximumCreditedIntervalSeconds: 60,
        nominalSampleDurationSeconds: 1
    )
}

public struct StressQualityDecision: Codable, Equatable, Sendable {
    public let accepted: Bool
    public let coveragePercentage: Double
    public let maximumGapSeconds: Int
    public let validDurationSeconds: Int
    public let rejectedDurationSeconds: Int
    public let windowDurationSeconds: Int
    public let baselineReady: Bool
    public let rejectionReasons: [StressQualityRejectionReason]
    public let diagnostics: [StressQualityDiagnostic]
    public let validSampleCount: Int
    public let rejectedSampleCount: Int
    public let duplicateTimestampCount: Int
    public let configurationVersion: String

    /// Twin of Kotlin `StressQualityDecision.withBaselineReadiness`.
    /// Attach causal baseline readiness without changing the signal evidence already measured for
    /// the window. Baseline rejection is always the final reason, preserving deterministic ordering.
    public func withBaselineReadiness(_ ready: Bool) -> StressQualityDecision {
        var reasons = rejectionReasons.filter { $0 != .baselineNotReady }
        if !ready { reasons.append(.baselineNotReady) }
        return StressQualityDecision(
            accepted: reasons.isEmpty, coveragePercentage: coveragePercentage,
            maximumGapSeconds: maximumGapSeconds, validDurationSeconds: validDurationSeconds,
            rejectedDurationSeconds: rejectedDurationSeconds, windowDurationSeconds: windowDurationSeconds,
            baselineReady: ready, rejectionReasons: reasons, diagnostics: diagnostics,
            validSampleCount: validSampleCount, rejectedSampleCount: rejectedSampleCount,
            duplicateTimestampCount: duplicateTimestampCount, configurationVersion: configurationVersion
        )
    }
}

public enum StressTemporalQuality {
    private struct CanonicalSample {
        let timestamp: Int
        let value: Double
        let durationSeconds: Int?
    }

    /// Twin of Kotlin `StressTemporalQuality.evaluate`.
    /// Analyze a half-open `[windowStart, windowEnd)` interval. Inputs are sorted and duplicate timestamps
    /// are collapsed before coverage or values are calculated, making replays deterministic.
    public static func evaluate(samples: [StressSignalSample],
                                windowStart: Int,
                                windowEnd: Int,
                                baselineReady: Bool,
                                configuration: StressQualityConfiguration = .daytimeV1)
        -> StressQualityDecision {
        let windowDelta = windowEnd.subtractingReportingOverflow(windowStart)
        let windowDuration = windowDelta.overflow ? 0 : max(0, windowDelta.partialValue)
        let inWindow = samples.filter { $0.timestamp >= windowStart && $0.timestamp < windowEnd }
        let outOfOrder = zip(inWindow, inWindow.dropFirst()).contains { pair in
            pair.0.timestamp > pair.1.timestamp
        }
        let duplicateCount = inWindow.count - Set(inWindow.map(\.timestamp)).count
        let canonical = canonicalValidSamples(inWindow)
        let invalid = inWindow.filter { !isValid($0) }

        let validIntervals = representedIntervals(canonical, windowStart: windowStart, windowEnd: windowEnd,
                                                 configuration: configuration)
        let rejectedIntervals = invalid.compactMap { sample -> Range<Int>? in
            guard windowDuration > 0 else { return nil }
            let duration = min(sample.durationSeconds.flatMap { $0 > 0 ? $0 : nil }
                                   ?? configuration.nominalSampleDurationSeconds,
                               configuration.maximumCreditedIntervalSeconds)
            let end = min(windowEnd, safeAdd(sample.timestamp, max(0, duration)))
            return end > sample.timestamp ? sample.timestamp..<end : nil
        }
        let validDuration = unionDuration(validIntervals)
        let rejectedDuration = unionDuration(rejectedIntervals)
        let maxGap = windowDuration > 0
            ? maximumGap(validIntervals, windowStart: windowStart, windowEnd: windowEnd) : 0
        let coverage = windowDuration > 0 ? Double(validDuration) / Double(windowDuration) : 0

        var reasons: [StressQualityRejectionReason] = []
        if windowDuration <= 0 { reasons.append(.invalidWindow) }
        if canonical.isEmpty { reasons.append(.noValidSamples) }
        if windowDuration > 0 && (validDuration < configuration.minimumValidDurationSeconds
            || coverage + 1e-12 < configuration.minimumCoverageFraction) {
            reasons.append(.insufficientTemporalCoverage)
        }
        if windowDuration > 0 && maxGap > configuration.maximumGapSeconds {
            reasons.append(.maximumGapExceeded)
        }
        if !baselineReady { reasons.append(.baselineNotReady) }

        var diagnostics: [StressQualityDiagnostic] = []
        if duplicateCount > 0 { diagnostics.append(.duplicateTimestamps) }
        if outOfOrder { diagnostics.append(.outOfOrderInput) }
        if !invalid.isEmpty { diagnostics.append(.rejectedSamples) }

        return StressQualityDecision(
            accepted: reasons.isEmpty,
            coveragePercentage: coverage * 100,
            maximumGapSeconds: maxGap,
            validDurationSeconds: validDuration,
            rejectedDurationSeconds: rejectedDuration,
            windowDurationSeconds: windowDuration,
            baselineReady: baselineReady,
            rejectionReasons: reasons,
            diagnostics: diagnostics,
            validSampleCount: canonical.count,
            rejectedSampleCount: invalid.count,
            duplicateTimestampCount: duplicateCount,
            configurationVersion: configuration.version
        )
    }

    /// Twin of Kotlin `StressTemporalQuality.canonicalMean`.
    /// Duration-weighted mean over canonical, non-overlapping represented intervals. This prevents a
    /// dense burst from outweighing a lower-cadence decoder that supplies honest interval durations.
    public static func canonicalMean(samples: [StressSignalSample], windowStart: Int, windowEnd: Int,
                                     configuration: StressQualityConfiguration = .daytimeV1) -> Double? {
        let canonical = canonicalValidSamples(
            samples.filter { $0.timestamp >= windowStart && $0.timestamp < windowEnd }
        )
        var weighted = 0.0
        var duration = 0
        for (index, sample) in canonical.enumerated() {
            let credit = min(sample.durationSeconds ?? configuration.nominalSampleDurationSeconds,
                             configuration.maximumCreditedIntervalSeconds)
            let next = index + 1 < canonical.count ? canonical[index + 1].timestamp : windowEnd
            let end = min(windowEnd, min(next, safeAdd(sample.timestamp, max(0, credit))))
            let seconds = max(0, end - max(windowStart, sample.timestamp))
            weighted += sample.value * Double(seconds)
            duration += seconds
        }
        return duration > 0 ? weighted / Double(duration) : nil
    }

    /// Twin of Kotlin `StressTemporalQuality.isValid`.
    private static func isValid(_ sample: StressSignalSample) -> Bool {
        guard sample.quality == .valid, !sample.isMissing,
              let value = sample.value, value.isFinite else { return false }
        if let duration = sample.durationSeconds, duration <= 0 { return false }
        return true
    }

    /// Twin of Kotlin `StressTemporalQuality.canonicalValidSamples`.
    private static func canonicalValidSamples(_ samples: [StressSignalSample]) -> [CanonicalSample] {
        let grouped = Dictionary(grouping: samples.filter(isValid), by: \.timestamp)
        return grouped.keys.sorted().compactMap { timestamp in
            guard let group = grouped[timestamp], !group.isEmpty else { return nil }
            let values = group.compactMap(\.value).sorted()
            guard !values.isEmpty else { return nil }
            let duration = group.compactMap(\.durationSeconds).max()
            return CanonicalSample(timestamp: timestamp,
                                   value: values.reduce(0, +) / Double(values.count),
                                   durationSeconds: duration)
        }
    }

    /// Twin of Kotlin `StressTemporalQuality.representedIntervals`.
    private static func representedIntervals(_ samples: [CanonicalSample],
                                             windowStart: Int,
                                             windowEnd: Int,
                                             configuration: StressQualityConfiguration) -> [Range<Int>] {
        guard windowEnd > windowStart else { return [] }
        return samples.enumerated().compactMap { index, sample in
            let duration = min(sample.durationSeconds ?? configuration.nominalSampleDurationSeconds,
                               configuration.maximumCreditedIntervalSeconds)
            let start = max(windowStart, sample.timestamp)
            let next = index + 1 < samples.count ? samples[index + 1].timestamp : windowEnd
            let end = min(windowEnd, min(next, safeAdd(sample.timestamp, max(0, duration))))
            return end > start ? start..<end : nil
        }
    }

    /// Twin of Kotlin `StressTemporalQuality.unionDuration`.
    private static func unionDuration(_ intervals: [Range<Int>]) -> Int {
        let ordered = intervals.sorted { $0.lowerBound < $1.lowerBound }
        guard var current = ordered.first else { return 0 }
        var total = 0
        for interval in ordered.dropFirst() {
            if interval.lowerBound <= current.upperBound {
                current = current.lowerBound..<max(current.upperBound, interval.upperBound)
            } else {
                total += current.count
                current = interval
            }
        }
        return total + current.count
    }

    /// Twin of Kotlin `StressTemporalQuality.maximumGap`.
    private static func maximumGap(_ intervals: [Range<Int>],
                                   windowStart: Int,
                                   windowEnd: Int) -> Int {
        guard windowEnd > windowStart else { return 0 }
        let ordered = intervals.sorted { $0.lowerBound < $1.lowerBound }
        guard var current = ordered.first else { return windowEnd - windowStart }
        var largest = max(0, current.lowerBound - windowStart)
        for interval in ordered.dropFirst() {
            if interval.lowerBound > current.upperBound {
                largest = max(largest, interval.lowerBound - current.upperBound)
                current = interval
            } else if interval.upperBound > current.upperBound {
                current = current.lowerBound..<interval.upperBound
            }
        }
        return max(largest, windowEnd - current.upperBound)
    }

    /// Twin of Kotlin `StressTemporalQuality.safeAdd`.
    private static func safeAdd(_ value: Int, _ nonnegativeDelta: Int) -> Int {
        let (sum, overflow) = value.addingReportingOverflow(nonnegativeDelta)
        return overflow ? Int.max : sum
    }
}
